### ISSUE TITLE
Gem Dependency Conflict

### ISSUE THREAD

**Issue by danapplegate** [fluent/fluent-plugin-s3#140]
Hi there,

I'm running into a gem dependency conflict when I try to use this plugin alongside the [`fluent-plugin-kinesis-firehose`](https://rubygems.org/gems/fluent-plugin-kinesis-firehose). It appears that fluent-plugin-kinesis-firehose requires `aws-sdk ~> 2.1.28`, which causes a conflict with the subsequently dependent `aws-sdk-resources` package from the s3 plugin's `aws-sdk ~> 2` dependency. Here's the exception I get during my Chef run:

```
---- Begin output of /etc/init.d/td-agent restart ----
           STDOUT: Restarting td-agent:  * td-agent
           STDERR: /opt/td-agent/embedded/lib/ruby/site_ruby/2.1.0/rubygems/specification.rb:2112:in `raise_if_conflicts': Unable to activate aws-sdk-2.1.36, because aws-sdk-resources-2.2.14 conflicts with aws-sdk-resources (= 2.1.36) (Gem::ConflictError)
            from /opt/td-agent/embedded/lib/ruby/site_ruby/2.1.0/rubygems/specification.rb:1280:in `activate'
            from /opt/td-agent/embedded/lib/ruby/site_ruby/2.1.0/rubygems/specification.rb:1314:in `block in activate_dependencies'
            from /opt/td-agent/embedded/lib/ruby/site_ruby/2.1.0/rubygems/specification.rb:1300:in `each'
            from /opt/td-agent/embedded/lib/ruby/site_ruby/2.1.0/rubygems/specification.rb:1300:in `activate_dependencies'
            from /opt/td-agent/embedded/lib/ruby/site_ruby/2.1.0/rubygems/specification.rb:1282:in `activate'
            from /opt/td-agent/embedded/lib/ruby/site_ruby/2.1.0/rubygems.rb:198:in `rescue in try_activate'
            from /opt/td-agent/embedded/lib/ruby/site_ruby/2.1.0/rubygems.rb:195:in `try_activate'
            from /opt/td-agent/embedded/lib/ruby/site_ruby/2.1.0/rubygems/core_ext/kernel_require.rb:126:in `rescue in require'

            from /opt/td-agent/embedded/lib/ruby/gems/2.1.0/gems/fluent-plugin-kinesis-firehose-0.1.1/lib/fluent/plugin/out_kinesis_firehose.rb:1:in `<top (required)>'
            from /opt/td-agent/embedded/lib/ruby/site_ruby/2.1.0/rubygems/core_ext/kernel_require.rb:54:in `require'
            from /opt/td-agent/embedded/lib/ruby/site_ruby/2.1.0/rubygems/core_ext/kernel_require.rb:54:in `require'
            from /opt/td-agent/embedded/lib/ruby/gems/2.1.0/gems/fluentd-0.12.20/lib/fluent/plugin.rb:168:in `block in try_load_plugin'
            from /opt/td-agent/embedded/lib/ruby/gems/2.1.0/gems/fluentd-0.12.20/lib/fluent/plugin.rb:166:in `each'
            from /opt/td-agent/embedded/lib/ruby/gems/2.1.0/gems/fluentd-0.12.20/lib/fluent/plugin.rb:166:in `try_load_plugin'
            from /opt/td-agent/embedded/lib/ruby/gems/2.1.0/gems/fluentd-0.12.20/lib/fluent/plugin.rb:126:in `new_impl'
            from /opt/td-agent/embedded/lib/ruby/gems/2.1.0/gems/fluentd-0.12.20/lib/fluent/plugin.rb:57:in `new_output'
            from /opt/td-agent/embedded/lib/ruby/gems/2.1.0/gems/fluentd-0.12.20/lib/fluent/agent.rb:127:in `add_match'
            from /opt/td-agent/embedded/lib/ruby/gems/2.1.0/gems/fluentd-0.12.20/lib/fluent/agent.rb:60:in `block in configure'
            from /opt/td-agent/embedded/lib/ruby/gems/2.1.0/gems/fluentd-0.12.20/lib/fluent/agent.rb:54:in `each'
            from /opt/td-agent/embedded/lib/ruby/gems/2.1.0/gems/fluentd-0.12.20/lib/fluent/agent.rb:54:in `configure'
            from /opt/td-agent/embedded/lib/ruby/gems/2.1.0/gems/fluentd-0.12.20/lib/fluent/root_agent.rb:82:in `configure'
            from /opt/td-agent/embedded/lib/ruby/gems/2.1.0/gems/fluentd-0.12.20/lib/fluent/engine.rb:117:in `configure'
            from /opt/td-agent/embedded/lib/ruby/gems/2.1.0/gems/fluentd-0.12.20/lib/fluent/engine.rb:91:in `run_configure'
            from /opt/td-agent/embedded/lib/ruby/gems/2.1.0/gems/fluentd-0.12.20/lib/fluent/supervisor.rb:515:in `run_configure'
            from /opt/td-agent/embedded/lib/ruby/gems/2.1.0/gems/fluentd-0.12.20/lib/fluent/supervisor.rb:186:in `dry_run'
            from /opt/td-agent/embedded/lib/ruby/gems/2.1.0/gems/fluentd-0.12.20/lib/fluent/supervisor.rb:133:in `start'
            from /opt/td-agent/embedded/lib/ruby/gems/2.1.0/gems/fluentd-0.12.20/lib/fluent/command/fluentd.rb:171:in `<top (required)>'
            from /opt/td-agent/embedded/lib/ruby/site_ruby/2.1.0/rubygems/core_ext/kernel_require.rb:54:in `require'
            from /opt/td-agent/embedded/lib/ruby/site_ruby/2.1.0/rubygems/core_ext/kernel_require.rb:54:in `require'
            from /opt/td-agent/embedded/lib/ruby/gems/2.1.0/gems/fluentd-0.12.20/bin/fluentd:6:in `<top (required)>'
            from /opt/td-agent/embedded/bin/fluentd:23:in `load'
            from /opt/td-agent/embedded/bin/fluentd:23:in `<top (required)>'
            from /usr/sbin/td-agent:7:in `load'
            from /usr/sbin/td-agent:7:in `<main>'
           ---- End output of /etc/init.d/td-agent restart ----
       Ran /etc/init.d/td-agent restart returned 1
```

To reproduce:
1. Use base Ubuntu 12.04
2. `sudo apt-get update`
3. Install td-agent `sudo apt-get install td-agent`
4. Install both td-agent-gems: `sudo td-agent-gem install fluent-plugin-s3 fluent-plugin-kinesis-firehose`
5. Edit `/etc/td-agent/td-agent.conf` to include match blocks with both plugins

```
<match *>
    type s3
    aws_key_id ...
    aws_sec_key ...
    s3_bucket ...
    s3_region us-east-1
    time_slice_format %Y%m%d
    buffer_path /var/log/td-agent/s3-email
    path logs/app/email
    utc
</match>
<match *>
    type kinesis_firehose
    delivery_stream_name ...
    aws_key_id ...
    aws_sec_key ...
    region us-east-1
    append_new_line true
</match>
```
1. `sudo service td-agent restart`, which fails with the above error and stack trace

Any ideas on how to resolve this issue?

**Comment by repeatedly** [comment]
I'm not sure why this problem happens because `~> 2` should cover all 2.x.x versions.

BTW, please check offcial gem, https://rubygems.org/gems/fluent-plugin-kinesis, instead.

**Comment by bratislavml** [comment]
Any news on this? @danapplegate did you solve this issue?

**Comment by gaohao** [comment]
looks like latest s3 and kinesis plugins have conflict on aws sdk cores

**Comment by kenhys** [comment]
It seems that:

* fluent-plugin-s3 requires aws-sdk-s3 ~> 1.60 then resolved to 1.96.2
* aws-sdk-s3 1.96.2 requires aws-sdk-core ~> 3, >= 3.112.0 

On the other hand,
* fluent-plugin-kinesis-firehose requires 'aws-sdk', '~> 2.1.28' then resolved to 2.1.36
* aws-sdk requires  aws-sdk-resources = 2.1.36 explicitly
* aws-sdk-resources requires aws-sdk-core = 2.1.36  exlicitly

As a result, they conflict. as mentioned at https://github.com/fluent/fluent-plugin-s3/issues/140#issuecomment-202616603, using official gem is better.
